### PR TITLE
Multi relay support

### PR DESCRIPTION
**Pull request by tompro**
## 📝 Description

Derive relay set from our nostr contacts with upper limit. We only connect to trusted and participant relays for now. Try to have at least one relay per contact (but try to fit all). Update relays on contact add. I need a bunch of relays to actually test all behaviors (especially the max total).

Relates to #613 

---

## ✅ Checklist

Please ensure the following tasks are completed before requesting a review:

- [x] My code adheres to the coding guidelines of this project.
- [x] I have run `cargo fmt`.
- [x] I have run `cargo clippy`.
- [x] I have added or updated tests (if applicable).
- [x] All CI/CD steps were successful.
- [x] I have updated the documentation (if applicable).
- [x] I have checked that there are no console errors or warnings.
- [x] I have verified that the application builds without errors.
- [x] I've described the changes made to the API. (modification, addition, deletion).

---

## 🚀 Changes Made

- **New Features:**
  - Active relay management

---

## 📋 Review Guidelines

Please focus on the following while reviewing:

- [ ] Does the code follow the repository's contribution guidelines?
- [ ] Are there any potential bugs or performance issues?
- [ ] Are there any typos or grammatical errors in the code or comments?

___

### **PR Type**
Enhancement

___

### **Description**
- Add dynamic relay management based on Nostr contacts
  - Implement relay calculation algorithm prioritizing trusted contacts
  - Add `get_all()` method to NostrContactStoreApi for fetching all contacts
  - Trigger relay refresh when contacts are added or updated

- Add `max_relays` configuration field with default of 50
  - Limit contact relays while always including user-configured relays

- Implement comprehensive relay calculation and update logic
  - Pass 1: Include all user relays (exempt from limit)
  - Pass 2: Add first relay from each trusted/participant contact
  - Pass 3: Fill remaining slots with additional contact relays

- Add extensive test coverage for relay calculation algorithm
  - 10 test cases covering edge cases and priority scenarios

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["NostrConfig<br/>max_relays field"] --> B["NostrClient<br/>refresh_relays()"]
  C["NostrContactStore<br/>get_all()"] --> B
  B --> D["calculate_relay_set_internal<br/>algorithm"]
  D --> E["update_relays()<br/>add/remove relays"]
  F["NostrContactProcessor<br/>upsert_contact()"] --> B
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>lib.rs</strong><dd><code>Add max_relays config field with default value</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-bfdb5500e043952ca4ce5c9602663e20f6adaf4f0d4e68d7b45a5e238c92cade">+14/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>lib.rs</strong><dd><code>Add nostr_max_relays config field to WASM Config</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-78b085a893e1faee301d123867503a432b3a173b3a527612442ed79927ceb308">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>mod.rs</strong><dd><code>Add get_all() to mock and update test config</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-49db059de3206a221445a6f9d2745edf8e81decc03c3ac08e49ceb1d727b2a02">+33/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mod.rs</strong><dd><code>Add get_all() to mock NostrContactStore implementation</code>&nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-43e5cc59ac540ede13a5547c11b72c4d6acc25b9f01d678ab32b2d1db9eea077">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_utils.rs</strong><dd><code>Add get_all() to mock and update test config</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-f4989df6c2d7b8bc93ea698f8e4a263222a663cf99af9e162827bacb4ca4241f">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>nostr.rs</strong><dd><code>Add get_all() method to NostrContactStoreApi trait</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-52ab08877fa9862f0169f65a12d8b21a4c01492f30cc584b5a8a1048040dd5b6">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>nostr_contact_store.rs</strong><dd><code>Implement get_all() to fetch all contacts from store</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-0670ccd712f28ee169d945125b594d179a1914cd0699f3f6294338b33d322f7c">+10/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>nostr_contact_processor.rs</strong><dd><code>Add relay refresh trigger on contact updates</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-8c996ef1b886539e34429453fd02e3675351c12b5215ae423ca7025a2d0f2040">+16/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>lib.rs</strong><dd><code>Pass contact store and trigger initial relay refresh</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-c714507fc66fabfa282a282f118fbdfa772f3c42cdcd1fc236e139830e6d37c9">+13/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>nostr.rs</strong><dd><code>Implement relay calculation and refresh logic with tests</code>&nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-8cfd79336972115e49e6641cef6699112af5ebf2b37e3249aa75db9c413fa595">+371/-9</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>context.rs</strong><dd><code>Pass contact store to create_nostr_clients function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Bitcredit-Core/pull/757/files#diff-4aab47a7683b01dc13d3bd22573728e6f581df241ee8a3e6eb19d12521bd635f">+7/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: crates/bcr-ebill-persistence/src/db/nostr_contact_store.rs [hunk 1]
```diff
@@ -65,6 +65,16 @@ impl NostrContactStoreApi for SurrealNostrContactStore {
...
     }
 
+    /// Get all Nostr contacts from the store.
+    async fn get_all(&self) -> Result<Vec<NostrContact>> {
+        let result: Vec<NostrContactDb> = self.db.select_all(Self::TABLE).await?;
+        let values = result
+            .into_iter()
+            .map(|c| c.to_owned().try_into().ok())
+            .collect::<Option<Vec<NostrContact>>>();
+        Ok(values.unwrap_or_default())
```
Comment (zupzup) [inline crates/bcr-ebill-persistence/src/db/nostr_contact_store.rs]: Hmm, doesn't this silently swallow the error, if e.g. all (or even 1) nostr contact in the DB would be invalid?

How about collecting into a `Result<Vec<NostrContact>>` and then propagating the error? I think even if this call would be "optional" in nature to succeed, this might help with debugging issues.

File: crates/bcr-ebill-api/src/lib.rs [hunk 2]
```diff
@@ -84,12 +84,25 @@ pub struct PaymentConfig {
...
+    /// Defaults to 50 if not specified.
+    pub max_relays: Option<usize>,
+}
+
+impl Default for NostrConfig {
+    fn default() -> Self {
+        Self {
+            only_known_contacts: false,
+            relays: vec![],
+            max_relays: Some(50),
```
Comment (codingpeanut157) [inline crates/bcr-ebill-api/src/lib.rs]:
```rust
Self {
   max_relays: Some(50),
  ..Default::default(),
}
```

**Review by zupzup — APPROVED** [review]
LGTM 👍

**Review by codingpeanut157 — APPROVED** [review]
minor comments left on the code

LGTM

File: crates/bcr-ebill-api/src/lib.rs (hunk 2 above)
Comment (tompro) [reply crates/bcr-ebill-api/src/lib.rs]: This would end up in a recursion I think.

File: crates/bcr-ebill-transport/src/nostr.rs [hunk 3]
```diff
@@ -1251,3 +1349,267 @@ mod tests {
...
+            .collect();
+
+    // Sort: Trusted (0) before Participant (1)
+    eligible_contacts.sort_by_key(|c| match c.trust_level {
+        TrustLevel::Trusted => 0,
+        TrustLevel::Participant => 1,
+        _ => 2, // unreachable due to filter
+    });
+
+    let contact_relay_limit = max_relays.unwrap_or(usize::MAX);
```
[the remark this answers is not in the document]
Comment (zupzup) [reply crates/bcr-ebill-transport/src/nostr.rs]: haha, yes :sweat_smile:
